File: library-core/src/main/java/com/mikepenz/fastadapter/utils/DefaultTypeInstanceCache.java
package com.mikepenz.fastadapter.utils;

import android.util.SparseArray;

import com.mikepenz.fastadapter.IItem;
import com.mikepenz.fastadapter.ITypeInstanceCache;

/**
 * Created by fabianterhorst on 24.08.17.
 */

public class DefaultTypeInstanceCache<Item extends IItem> implements ITypeInstanceCache<Item> {

    // we remember all possible types so we can create a new view efficiently
    private final SparseArray<Item> mTypeInstances = new SparseArray<>();

    @Override
    public boolean register(Item item) {
        if (mTypeInstances.indexOfKey(item.getType()) < 0) {
            mTypeInstances.put(item.getType(), item);
            return true;
        }
        return false;
    }

    @Override
    public Item get(int type) {
        return mTypeInstances.get(type);
    }

    @Override
    public void clear() {
        mTypeInstances.clear();
    }
}

File: library-core/src/main/java/com/mikepenz/fastadapter/utils/DefaultTypeInstanceCache.kt
package com.mikepenz.fastadapter.utils

import android.util.SparseArray
import com.mikepenz.fastadapter.GenericItem
import com.mikepenz.fastadapter.ITypeInstanceCache

/**
 * Created by fabianterhorst on 24.08.17.
 */
class DefaultTypeInstanceCache<Item : GenericItem> :
        ITypeInstanceCache<Item> {

    // we remember all possible types so we can create a new view efficiently
    private val typeInstances = SparseArray<Item>()

    override fun register(item: Item): Boolean {
        if (typeInstances.indexOfKey(item.type) < 0) {
            typeInstances.put(item.type, item)
            return true
        }
        return false
    }

    override fun get(type: Int): Item {
        return typeInstances.get(type)
    }

    override fun clear() {
        typeInstances.clear()
    }
}
